keps/sig-storage/20191125-bucket-provisioning.md

```diff
@@ -0,0 +1,540 @@
+  - "@erinboyd"
+  - "@guymguym"
+  - "@travisn"
+approvers:
+  - TBD
+editor: TBD
+creation-date: 2019-11-25
+last-updated: 2019-11-15
+status: implementable
+see-also:
+  - "https://github.com/kube-object-storage/lib-bucket-provisioner"
+  - "https://github.com/kube-object-storage/lib-bucket-provisioner/blob/master/doc/design/object-bucket-lib.md"
+  - "https://github.com/kube-object-storage/lib-bucket-provisioner/tree/master/doc/examples"
+---
+
+# Object Bucket Provisioning
+
+
+## Table of Contents
+
+<!-- toc -->
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+- [Design Details](#design-details)
+  - [Custom Resources](#custom-resources)
+  - [Moving Parts](#moving-parts)
+  - [Interfaces](#interfaces)
+  - [Library Usage](#library-usage)
+  - [Current Restrictions](#current-restrictions)
+  - [Risks and Mitigations](#risks-and-mitigations)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+    - [Examples](#examples)
+      - [Alpha -&gt; Beta Graduation](#alpha---beta)
+      - [Beta -&gt; GA Graduation](#beta---ga)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Version Skew Strategy](#version-skew-strategy)
+- [Implementation History](#implementation-history)
+- [Drawbacks [optional]](#drawbacks-optional)
+- [Alternatives [optional]](#alternatives-optional)
+<!-- /toc -->
+
+## Summary
+
+Kubernetes natively supports dynamic provisioning for file and block storage but lacks support for object bucket provisioning.
+Without a formal API to describe and manage bucket provisioning, storage providers are left to write their own provisioners including all of the necessary controller code.
+This "wild west" landscape for bucket provisioning is potentially confusing, overly complex, inconsistent, may slow down Kubernetes adoption by object store users, and may hinder application portability.
+
+We propose a common Kubernetes control-plane API for the management of object store bucket lifecycles which has no dependencies on the underlying providers.
+
+## Motivation
+
+By defining a Kubernetes API for object bucket provsioning, we provide a consistent framework for all object store providers and thus enhance the end user's storage experience. This API allows object store vendors to focus on how they wish to provision buckets instead of writing and testing the controller aspects of provisioning.
+
+### Goals
+
++ Define a _control-plane_ object bucket management API thus relieving object store providers from Kubernetes controller details.
++ Minimize technical ramp-up for storage vendors who have already contributed external storage provisioners (file and block).
++ Make bucket consumption similar to file or block consumption using familiar concepts and commands.
++ Use native Kubernetes resources where possible, again, to keep the bucket experience for users and admins similar to existing storage provisioning.
```

Comment:
I'm assuming this refers to the decision to overload StorageClass for parameters.  I do not see the value in this - I see an anti-value.  Volume provisioning is fundamentally different than provisioning some external entity.  Artificial association is more likely to confuse IMO than reduce confusion.

Reply:
**This concern should be discussed**.
My team felt/feels it is important to use native storage classes for the reason cited. We know others who use a custom resource, similar to a storage class, but specific to their object store needs.

Reply:
There are several benefits to a separate CR:

  * Avoiding cognitive dissonance - StorageClass is used for volume operations.  Other classes (e.g., VolumeSnapshotClass) have been introduced for other class-style parameterization.  Adding ObjectBucketClass would seem to be the most logical approach.  

  * You can evolve it independently of StorageClass, which responds to requirements from block/file (you don't have to bring them along, and you don't have to be brought along).

  * StorageClass is non-namespaced.  You've made progress towards making the object support install-able and usable without requiring write access to non-namespace objects.  If you had a single well-known namespace for object operations, ObjectBucketClass resources could go into that namespace (along with provisioners and OBs).

Reply:
These are all pretty compelling points for a bespoke BucketClass object.  We leaned on SCs initially because of assumptions we made about the system being familiar to k8s users.  However, like you point out, we're beholden to the API which is exclusively intended for file/block.  Namespacing the classes with their provisioners also lends itself to scalability.